### PR TITLE
feat: Add support for `IO[bytes]` and `bytes` in `scan_{...}` functions

### LINKED ISSUE #4950: Add BytesIO support to `scan_csv`
### Problem Description

Firstly, thank you for making this fantastic library 😀 

I have the following use case:

```
import zstandard
import polars as pl

with open(path, 'rb') as f:
    dctx = zstandard.ZstdDecompressor()
        with dctx.stream_reader(f) as reader:
            df = pl.read_csv(reader, sep='\t')
```

Where `path` is the path of a Zstandard compressed TSV file. I'm working with bioinformatics data, and bioinformaticians love to generate massive CSVs/TSVs and then compress them.

I would like to use `scan_csv` to read the decompressed BytesIO stream instead and take advantage of all the cool lazy evaluation features to reduce memory usage. Alternatively, it would be great if `scan_csv` supported Zstandard compressed file paths directly. 

Thanks for your time!

### LINKED ISSUE #12617: Support BytesIO, StringIO etc. in scan_csv()
### Description

`pl.read_csv(StringIO('a\n1'))` works
`pl.read_csv(BytesIO(b'a\n1'))` works

But:

1. `pl.scan_csv(StringIO('a\n1')).collect()` gives:
```
polars.exceptions.ComputeError: error while reading a
: No such file or directory (os error 2): a
```
2. `pl.scan_csv(BytesIO(b'a\n1')).collect()` gives: 
```
TypeError: argument 'paths': 'bytes' object cannot be converted to 'PyString'
```
3. `pl.scan_csv([StringIO('a\n1'), StringIO('a\n2')]).collect()` gives:
```
TypeError: expected str, bytes or os.PathLike object, not StringIO
```
4. `pl.scan_csv([BytesIO(b'a\n1'), BytesIO(b'a\n2')]).collect()` gives:
```
TypeError: expected str, bytes or os.PathLike object, not BytesIO
```

### LINKED ISSUE #18581: Incorrect/mismatched LazyFrame schema resolution between `collect` and `collect_schema`
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from tempfile import NamedTemporaryFile

import polars as pl


ctx = pl.SQLContext()


with NamedTemporaryFile() as f:
    f.write(b"a,b,c\n1,2,3\n4,5,6")
    f.flush()

    ctx.register("test", pl.scan_csv(f.name))
    lf = ctx.execute("SELECT COUNT(*) AS x FROM test")

    lazy_schema = lf.collect_schema()
    actual_schema = lf.collect().schema


expected = pl.Schema({"x": pl.UInt32})


# The lazy schema is correct
assert lazy_schema == expected
# The actual schema is incorrect (column name is `len`, not `x`)
assert actual_schema == expected

```

### Log output

```shell
$ POLARS_VERBOSE=1 python bug.py
Traceback (most recent call last):
  File "/home/jcristharif/Code/ibis/bug.py", line 26, in <module>
    assert actual_schema == expected
           ^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

### Issue description

Sometimes when executing SQL against a lazyframe the resulting column names don't match (the alias is ignored). If I swap out the `scan_csv` for a `read_csv` (or an in-memory `DataFrame`) things work properly.

### Expected behavior

SQL aliases to be properly respected.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.6.0
Index type:          UInt32
Platform:            Linux-6.5.0-45-generic-x86_64-with-glibc2.35
Python:              3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               5.4.0
cloudpickle          3.0.0
connectorx           <not installed>
deltalake            0.18.2
fastexcel            <not installed>
fsspec               2024.6.1
gevent               <not installed>
great_tables         <not installed>
matplotlib           3.9.1
nest_asyncio         1.6.0
numpy                2.1.1
openpyxl             <not installed>
pandas               2.2.2
pyarrow              17.0.0
pydantic             2.8.2
pyiceberg            <not installed>
sqlalchemy           2.0.32
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by coastalwhite**
This PR adds support for the `bytes` and `IO[bytes]` interface to all `scan_{parquet, csv, ...}` functions.

Since the same code was being touched. This PR also fixes #18581.

Fixes #4950.
Fixes #12617.

### CODE REVIEW

File: crates/polars-mem-engine/src/executors/scan/csv.rs [hunk 1]
```diff
@@ -56,56 +57,49 @@ impl CsvExec {
...
-                            // Safety: This was initialized by schema inference.
-                            .unwrap()
-                            .try_open_assume_latest()?;
-                        let owned = &mut vec![];
+                let source = self.sources.at(i);
+                let owned = &mut vec![];
+
+                let memslice = match source {
+                    ScanSourceRef::File(path) => {
+                        let file = if run_async {
```
Comment (coastalwhite) [inline crates/polars-mem-engine/src/executors/scan/csv.rs:73]: drive-by: `feature_gated` is preferred

File: crates/polars-mem-engine/src/executors/scan/ipc.rs [hunk 2]
```diff
@@ -22,23 +21,20 @@ pub struct IpcExec {
...
 
         let mut out = if is_cloud || force_async {
-            #[cfg(not(feature = "cloud"))]
-            {
-                panic!("activate cloud feature")
-            }
-
-            #[cfg(feature = "cloud")]
-            {
+            feature_gated!("cloud", {
```
Comment (coastalwhite) [inline crates/polars-mem-engine/src/executors/scan/ipc.rs:31]: drive-by: `feature_gated` is preferred

File: crates/polars-mem-engine/src/utils.rs [hunk 3]
```diff
@@ -1,22 +1,28 @@
...
 use polars_utils::aliases::PlHashSet;
 use polars_utils::arena::{Arena, Node};
 
 /// Get a set of the data source paths in this LogicalPlan
-pub(crate) fn agg_source_paths(
+///
+/// # Notes
+///
+/// - Scan sources with in-memory buffers are ignored.
+pub(crate) fn agg_source_paths<'a>(
```
Comment (coastalwhite) [inline crates/polars-mem-engine/src/utils.rs:13]: drive-by: remove the allocation here

**Comment by coastalwhite** [comment]
Putting this back into draft for a second. I want to try to remove the specialized `read_{_}` impls as well with this PR now so I made it possible to pass opened files. This did create some further problems, which I still need to fix.

**Comment by coastalwhite** [comment]
> Putting this back into draft for a second. I want to try to remove the specialized `read_{_}` impls as well with this PR now so I made it possible to pass opened files. This did create some further problems, which I still need to fix.

In the end I chose not to do this, because it would require significant mappings.

File: py-polars/tests/unit/io/test_ipc.py [hunk 4]
```diff
@@ -339,29 +340,3 @@ def test_ipc_decimal_15920(
         path = f"{tmp_path}/data"
         df.write_ipc(path)
         assert_frame_equal(pl.read_ipc(path), df)
-
```
Comment (coastalwhite) [inline py-polars/tests/unit/io/test_ipc.py:342]: I removed this test as it does not really make much sense. The mmaps are unmapped when they are finished, I feel like a bug was triggering this behavior before. For the streaming engine it does make a lot of sense to have a test though and the checks for this are still in place.

**Comment by coastalwhite** [comment]
This solves one of the subgoals of #13040.

**Comment by coastalwhite** [comment]
In the end, a lot more things had to be touched than I thought originally. @ritchie46 sorry for the large review :sweat:

File: crates/polars-io/src/csv/read/parser.rs [hunk 5]
```diff
@@ -18,26 +19,21 @@ use crate::utils::maybe_decompress_bytes;
...
     separator: u8,
     quote_char: Option<u8>,
     comment_prefix: Option<&CommentPrefix>,
     eol_char: u8,
     has_header: bool,
 ) -> PolarsResult<usize> {
     let file = if is_cloud_url(path) || config::force_async() {
-        #[cfg(feature = "cloud")]
-        {
+        feature_gated!("cloud", {
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-io/src/csv/read/parser.rs:30]: Yes, much nicer.

**Review by ritchie46 — APPROVED** [review]

File: crates/polars-io/src/ipc/ipc_file.rs [hunk 6]
```diff
@@ -52,6 +52,7 @@ use crate::RowIndex;
 #[derive(Clone, Debug, PartialEq, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct IpcScanOptions {
+    /// Not used anymore.
```
[the remark this answers is not in the document]
Comment (coastalwhite) [reply crates/polars-io/src/ipc/ipc_file.rs:55]: Yeah we can. It is now just always MMapped.

[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-io/src/ipc/ipc_file.rs:55]: Alright, can you remove that, then we can merge. :)
